cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/RemoteAgentService.java

```diff
@@ -0,0 +1,164 @@
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+/**
+ * The main class to run the remote agent service.
+ */
+public class RemoteAgentService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteAgentService.class);
+  private static final Gson GSON = new Gson();
+  public static final String CONNECT_CONTROL_CHANNEL = "/v3/tethering/controlchannels/";
+
+  private final CConfiguration cConf;
+  private final TetherStore store;
+  private final String instanceName;
+
+  @Inject
+  public RemoteAgentService(CConfiguration cConf, TransactionRunner transactionRunner) {
+    this.cConf = cConf;
+    this.store = new TetherStore(transactionRunner);
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  @Override
+  public void startUp() throws InstantiationException, IllegalAccessException {
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tether.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+    }
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    int connectionInterval = cConf.getInt(Constants.Tether.CONNECT_INTERVAL,
+                                          Constants.Tether.CONNECT_INTERVAL_DEFAULT);
+    // Try right away if there's anything to cleanup, we will then schedule based on the minimum retention interval
+    return Scheduler.newFixedRateSchedule(connectionInterval, connectionInterval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void runOneIteration() {
+    List<PeerInfo> peers;
+    try {
+      peers = store.getPeers().stream()
+        // Ignore peers in REJECTED state.
+        .filter(p -> p.getTetherStatus() != TetherStatus.REJECTED)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return;
+    }
+    for (PeerInfo peer : peers) {
+      try {
+        HttpResponse resp = TetherUtils.sendHttpRequest(HttpMethod.GET, new URI(peer.getEndpoint())
+          .resolve(CONNECT_CONTROL_CHANNEL + instanceName));
+        switch (resp.getResponseCode()) {
+          case HttpURLConnection.HTTP_OK:
+            TetherStatus peerStatus = store.getTetherStatus(peer.getName());
```

Comment:
Why need to lookup again? Isn't the status in the `PeerInfo` already?

Reply:
fixed